sys/usb/usbus/msc/msc.c

```diff
@@ -0,0 +1,284 @@
+extern mtd_dev_t *mtd0;
+
+/* Internal buffer to handle size difference between MTD layer and USB
+   endpoint size as some MTD implementation (like mtd_sdcard) doesn't allow
+   endpoint size transfer */
+static unsigned char buff[USBUS_MSC_BUFFER_SIZE];
+
+/* Internal handler definitions */
+static void _event_handler(usbus_t *usbus, usbus_handler_t *handler,
+                           usbus_event_usb_t event);
+
+static int _control_handler(usbus_t *usbus, usbus_handler_t *handler,
+                            usbus_control_request_state_t state,
+                            usb_setup_t *setup);
+
+static void _transfer_handler(usbus_t *usbus, usbus_handler_t *handler,
+                              usbdev_ep_t *ep, usbus_event_transfer_t event);
+
+static void _init(usbus_t *usbus, usbus_handler_t *handler);
+
+static const usbus_handler_driver_t msc_driver = {
+    .init = _init,
+    .event_handler = _event_handler,
+    .control_handler = _control_handler,
+    .transfer_handler = _transfer_handler,
+};
+
+ static const usbus_descr_gen_funcs_t _msc_descriptor = {
+    .fmt_post_descriptor = NULL,
+    .fmt_pre_descriptor = NULL,
+    .len = {
+        .fixed_len =  0,
+    },
+    .len_type = USBUS_DESCR_LEN_FIXED,
+};
+
+static void _write_xfer(usbus_msc_device_t *msc) {
+/* Check if we have a block to read and transfer */
+    if (msc->block_nb) {
+        size_t len;
+        /* Retrieve incoming data */
+        usbdev_ep_get(msc->ep_out->ep, USBOPT_EP_AVAILABLE, &len, sizeof(size_t));
+        if (len > 0) {
+            /* Prepare page buffer */
+            memcpy(&msc->buffer[msc->block_offset], msc->ep_out->ep->buf, len);
+        }
+        usbdev_ep_ready(msc->ep_out->ep, 0);
+        /* Update offset for page buffer */
+        msc->block_offset += len;
+        /* Decrement whole len */
+        msc->cmd.len -= len;
+
+        /* buffer is full, write it and point to new block if any */
+        if (msc->block_offset >= (mtd0->page_size * msc->pages_per_vpage)) {
+            mtd_write_page(mtd0, msc->buffer, msc->block * msc->pages_per_vpage,
+                           0, mtd0->page_size * msc->pages_per_vpage);
+            msc->block_offset = 0;
+            msc->block++;
+            msc->block_nb--;
+        }
+        if (msc->cmd.len == 0) {
+            /* All blocks have been transferred, send CSW to host */
+            if (msc->state == DATA_TRANSFER) {
+                msc->state = GEN_CSW;
+            }
+        }
+    }
+}
+static void _xfer_data( usbus_msc_device_t *msc)
+{
+    /* Check if we have a block to read and transfer */
+    if (msc->block_nb) {
+        /* read buffer from mtd device */
+        if (msc->block_offset == 0) {
+            mtd_read_page(mtd0, msc->buffer, msc->block * msc->pages_per_vpage,
+                          0, mtd0->page_size * msc->pages_per_vpage);
+        }
+        /* Prepare endpoint buffer */
+        memcpy(msc->ep_in->ep->buf, &msc->buffer[msc->block_offset], 64);
```

Comment:
Why 64?

Reply:
I will replace 64 by ep->len.
See the other comment for more details.

Reply:
Updated.